data/txDupCache.go

```diff
@@ -191,51 +192,74 @@ func (c *txSaltedCache) innerCheck(msg []byte) (*crypto.Digest, bool) {
 
 	d := crypto.Digest(blake2b.Sum256(toBeHashed))
 
-	_, found := c.cur[d]
+	v, found := c.cur[d]
 	if found {
-		return nil, true
+		return &d, v, &c.cur, true
 	}
 
 	toBeHashed = append(toBeHashed[:len(msg)], c.prevSalt[:]...)
 	toBeHashed = toBeHashed[:len(msg)+len(c.prevSalt)]
 	pd := crypto.Digest(blake2b.Sum256(toBeHashed))
-	_, found = c.prev[pd]
+	v, found = c.prev[pd]
 	if found {
-		return nil, true
+		return &pd, v, &c.prev, true
 	}
-	return &d, false
+	return &d, nil, nil, false
 }
 
 // CheckAndPut adds msg into a cache if not found
 // returns a hashing key used for insertion if the message not found.
-func (c *txSaltedCache) CheckAndPut(msg []byte) (*crypto.Digest, bool) {
+func (c *txSaltedCache) CheckAndPut(msg []byte, sender network.Peer) (*crypto.Digest, *sync.Map, bool) {
 	c.mu.RLock()
-	d, found := c.innerCheck(msg)
+	d, vals, page, found := c.innerCheck(msg)
 	salt := c.curSalt
-	c.mu.RUnlock()
 	// fast read-only path: assuming most messages are duplicates, hash msg and check cache
+	// keep lock - it is needed for copying vals in defer
+	senderFound := false
 	if found {
-		return d, found
+		if _, senderFound = vals.Load(sender); senderFound {
+			c.mu.RUnlock()
+			return d, vals, true
+		}
 	}
+	c.mu.RUnlock()
```

Comment:
Do you need a write lock for all the cases below?
If only the sender is missing, it can be added without a write lock right?

Reply:
this is valid, I'll refactor the code. Thank you. I did not update locking after switching from map to sync.Map

Reply:
pushed an updated